File: src/main/java/com/zynchub/digital/hubservice/app/feature/flags/mapper/FeatureMapper.java
package com.zynchub.digital.hubservice.app.feature.flags.mapper;

import com.zynchub.digital.hubservice.app.feature.flags.dto.FeatureDto;
import com.zynchub.digital.hubservice.app.feature.flags.model.FeatureFlag;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public class FeatureMapper {

    public List<FeatureFlag> mapFromDto(List<FeatureDto> featureDtos) {
        return featureDtos.stream().map(this::mapFeature).toList();
    }

    private FeatureFlag mapFeature(FeatureDto featureDto) {
        FeatureFlag feature = new FeatureFlag();
        feature.setName(featureDto.name());
        feature.setParent(featureDto.parent() == null || featureDto.parent().isBlank()
                ? featureDto.name()
                : featureDto.parent());
        feature.setEnabled(featureDto.enabled());
        feature.setDescription(featureDto.description());
        return feature;
    }
}

File: internal/app/mapper/feature_mapper.go
package mapper

import (
	"plexus-bff-service-go/internal/app/dto"
	"plexus-bff-service-go/internal/app/model"
)

type FeatureMapper struct{}

func NewFeatureMapper() *FeatureMapper {
	return &FeatureMapper{}
}

func (m *FeatureMapper) MapFromDTO(features []dto.FeatureFlag) []model.FeatureFlag {
	result := make([]model.FeatureFlag, 0, len(features))
	for _, feature := range features {
		parent := feature.Parent
		if parent == "" {
			parent = feature.Name
		}
		result = append(result, model.FeatureFlag{
			Name:        feature.Name,
			Parent:      parent,
			Enabled:     feature.Enabled,
			Description: feature.Description,
		})
	}
	return result
}
